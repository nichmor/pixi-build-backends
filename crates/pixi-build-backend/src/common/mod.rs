//! Common utilities that are shared between the different build backends.
mod configuration;
mod requirements;
mod variants;

pub use configuration::{build_configuration, BuildConfigurationParams};
pub use requirements::requirements;
pub use variants::compute_variants;
