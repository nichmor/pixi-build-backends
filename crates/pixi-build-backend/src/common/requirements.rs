use std::collections::BTreeMap;

use rattler_build::{
    recipe::{parser::Requirements, variable::Variable},
    NormalizedKey,
};
use rattler_conda_types::ChannelConfig;

use crate::{dependencies::extract_dependencies, traits::Dependencies, ProjectModel, Targets};

/// Return requirements for the given project model
pub fn requirements<P: ProjectModel>(
    dependencies: Dependencies<<P::Targets as Targets>::Spec>,
    channel_config: &ChannelConfig,
    variant: &BTreeMap<NormalizedKey, Variable>,
) -> miette::Result<Requirements> {
    // Extract dependencies into requirements
    let requirements = Requirements {
        build: extract_dependencies(channel_config, dependencies.build, variant)?,
        host: extract_dependencies(channel_config, dependencies.host, variant)?,
        run: extract_dependencies(channel_config, dependencies.run, variant)?,
        ..Default::default()
    };
    Ok(requirements)
}
