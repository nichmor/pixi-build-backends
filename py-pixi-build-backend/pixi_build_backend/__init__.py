"""
Python implementation of pixi-build-python backend using Python bindings.
"""

# __all__ = [
#     "PythonGenerator",
#     "PythonBackendConfig",
#     "BuildScriptContext",
#     "Installer",
#     "BuildPlatform",
#     "ProjectModelV1",
# ]
