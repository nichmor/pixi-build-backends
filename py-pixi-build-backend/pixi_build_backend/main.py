"""
Main entry point for Python backend implementation.
"""

import sys
import asyncio
from pixi_build_backend.types.generated_recipe import GenerateRecipeProtocol
from pixi_build_backend.pixi_build_backend import py_main, PyGenerateRecipe


async def main_entry_point(instance: GenerateRecipeProtocol):
    """Main entry point for the build backend"""
    py_generator = PyGenerateRecipe(instance)

    # Remove python name from argv
    args = sys.argv[1:] if len(sys.argv) > 1 else []
    print("sys argv", sys.argv)

    try:
        await py_main(py_generator, sys.argv)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
