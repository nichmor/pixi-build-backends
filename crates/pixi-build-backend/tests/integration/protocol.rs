use crate::common::model::{convert_test_model_to_project_model_v1, load_project_model_from_json};
use pixi_build_backend::{
    generated_recipe::GeneratedRecipe,
    intermediate_backend::{IntermediateBackend, IntermediateBackendConfig},
    protocol::Protocol,
};
use pixi_build_types::{
    ChannelConfiguration, PlatformAndVirtualPackages,
    procedures::{conda_build::CondaBuildParams, conda_metadata::CondaMetadataParams},
};
use rattler_build::console_utils::LoggingOutputHandler;
use rattler_conda_types::Platform;
use tempfile::TempDir;
use url::Url;

#[tokio::test]
async fn test_conda_get_metadata() {
    let tmp_dir = TempDir::new().unwrap();
    let tmp_dir_path = tmp_dir.path().to_path_buf();

    let pixi_manifest = tmp_dir_path.join("pixi.toml");

    let build_dir = tmp_dir_path.join("build");

    // Load a model from JSON
    let original_model = load_project_model_from_json("minimal_project_model.json");

    // Serialize it back to JSON
    let project_model_v1 = convert_test_model_to_project_model_v1(original_model);

    // save the pixi.toml file to a temporary location
    fs_err::write(&pixi_manifest, toml::to_string(&project_model_v1).unwrap()).unwrap();

    // Convert to IR
    let generated_recipe = GeneratedRecipe::from_model(project_model_v1, pixi_manifest.clone());

    let platform = PlatformAndVirtualPackages {
        platform: Platform::current(),
        virtual_packages: None,
    };

    let channel_configuration = ChannelConfiguration {
        base_url: Url::parse("https://prefix.dev").unwrap(),
    };

    let params = CondaMetadataParams {
        build_platform: Some(platform.clone()),
        host_platform: Some(platform.clone()),
        channel_base_urls: Some(vec![]),
        channel_configuration,
        variant_configuration: None,
        work_directory: build_dir,
    };

    let intermediate_backend = IntermediateBackend::new(
        pixi_manifest.clone(),
        generated_recipe,
        IntermediateBackendConfig::default(),
        LoggingOutputHandler::default(),
        None,
    )
    .unwrap();

    let conda_metadata = intermediate_backend
        .conda_get_metadata(params)
        .await
        .unwrap();

    insta::assert_yaml_snapshot!(conda_metadata, {
        ".packages[0].sources.boltons.path" => "[redacted]",
        ".packages[0].subdir" => "[redacted]",
    })
}

#[tokio::test]
async fn test_conda_build() {
    let tmp_dir = TempDir::new().unwrap();
    let tmp_dir_path = tmp_dir.path().to_path_buf();

    let pixi_manifest = tmp_dir_path.join("pixi.toml");
    let build_dir = tmp_dir_path.join("build");

    // Load a model from JSON
    let original_model = load_project_model_from_json("minimal_project_model_for_build.json");

    // Serialize it back to JSON
    let project_model_v1 = convert_test_model_to_project_model_v1(original_model);

    // save the pixi.toml file to a temporary location
    fs_err::write(&pixi_manifest, toml::to_string(&project_model_v1).unwrap()).unwrap();

    // Convert to IR
    let mut generated_recipe = GeneratedRecipe::from_model(project_model_v1, pixi_manifest.clone());

    generated_recipe.recipe.build.script = vec!["echo 'Hello, World!'".to_string()];

    let channel_configuration = ChannelConfiguration {
        base_url: Url::parse("https://prefix.dev").unwrap(),
    };

    let build_params = CondaBuildParams {
        build_platform_virtual_packages: None,
        host_platform: None,
        channel_base_urls: None,
        channel_configuration,
        outputs: None,
        variant_configuration: None,
        work_directory: build_dir.clone(),
        editable: false,
    };

    let intermediate_backend = IntermediateBackend::new(
        pixi_manifest.clone(),
        generated_recipe,
        IntermediateBackendConfig::default(),
        LoggingOutputHandler::default(),
        None,
    )
    .unwrap();

    let conda_build_result = intermediate_backend
        .conda_build(build_params)
        .await
        .unwrap();

    insta::assert_yaml_snapshot!(conda_build_result, {
        ".packages[0].output_file" => "[redacted]",
        ".packages[0].subdir" => "[redacted]",
    });
}
