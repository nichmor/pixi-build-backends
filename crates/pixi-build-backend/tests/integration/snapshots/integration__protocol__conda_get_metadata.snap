---
source: crates/pixi-build-backend/tests/integration/protocol.rs
expression: conda_metadata
---
packages:
  - name: minimal-package
    version: 1.0.0
    build: hbf21a9e_0
    buildNumber: 0
    subdir: osx-arm64
    depends:
      - boltons
      - python
    constraints: []
    license: ~
    licenseFamily: ~
    noarch: false
    sources:
      boltons:
        Path:
          path: "[redacted]"
inputGlobs: ~
